support keeping parameters constant during the calibration

dblaschke-LANL opened the pull request:
sometimes we want to keep some parameters fixed while calibrating others; this was not previously possible but is now implemented with this patch: users may (optionally) pass some paramaters via the constants variable to supercal, the code will no longer crash in this case, but rather keep those parameters constant and vary only the others.
Edits in several places were needed:
.) the initialize() method of the MaterialModel class needs to know to look for missing parameters in the constants
.) the constraints functions in supercal need an additional input variable (the constants) in order to find the values of ‘constant’ parameters; .) because of the above, the checkConstraints() method of the CalibSetup() class now accepts an additional input variable ‘constants’ and calls to this method were updated accordingly throughout the calibration code
.) the constants are received from the ModelMaterialStrength class as an attribute and in order to have a consistent api, dummy attributes .constants=None were added to the other models in models_withLik.py
.) since supercal is now using the constraint fct of ptw defined in physics, it makes sense to import both submodules upon “import impala” (see additional line in __init__.py)
.) some minor code cleanup, mostly in the ModelParameters class

dblaschke-LANL committed:
support keeping parameters constant during the calibration

sometimes we want to keep some parameters fixed while calibrating others; this was not
previously possible but is now implemented with this patch:
users may (optionally) pass some paramaters via the constants variable to supercal,
the code will no longer crash in this case, but rather keep those parameters constant
and vary only the others.
Edits in several places were needed:
.) the initialize() method of the MaterialModel class needs to know to look for missing
parameters in the constants
.) the constraints functions in supercal need an additional input variable (the constants)
in order to find the values of ‘constant’ parameters;
.) because of the above, the checkConstraints() method of the CalibSetup() class now accepts
an additional input variable ‘constants’ and calls to this method were updated accordingly
throughout the calibration code
.) the constants are received from the ModelMaterialStrength class as an attribute and in order to have a consistent api, dummy attributes .constants=None were added to the other models in models_withLik.py
.) since supercal is now using the constraint fct of ptw defined in physics, it makes sense to import both submodules upon “import impala” (see additional line in __init__.py)
.) some minor code cleanup, mostly in the ModelParameters class

dblaschke-LANL committed:
complete the patch:

chol_sample_nper_constraints() and chol_sample_1per_constraints()
    require one additional input variable ('constants')

dblaschke-LANL committed:
maintain backwards compatibility in CheckConstraints

simplify CheckConstraints api while maintaining backwards compatibility

luiarthur committed:
Merge branch 'master' into support_const_params

luiarthur committed:
format __init__.py

luiarthur committed:
format physical_models_vec

luiarthur committed:
format all changed files

luiarthur committed:
lint

luiarthur commented:
> 3\. I re-ran some PTW calibrations with the new code and different combinations of keeping some PTW parameters fixed at values that the first run (with all of them free) had calibrated them to and got consistent results.

That's excellent! Are you able to add this into our `tests/`?

> We can think about additional tests to add - probably some calibration test with PTW - but that would likely be a lot slower than the friedman test.

That would be valuable, but I'll leave that up to you. I'm inclined to say let's leave this for later.

dblaschke-LANL committed:
move beta from consts into ptw params

.) promoting beta to a ptw parameter serves 2 purposes:
   - allows the user to calibrate it (previously was always const)
   - and since our tests still initialize beta as a const, part of
   this merge request is thereby automatically tested
.) also remove some obsolete code:
   - fct fast_pow() was useful in ancient numpy versions, but
     nowadays np.power() is faster
   - no need for 'try - except' constructs around assert statements

dblaschke-LANL committed:
add a new test for pooled calibration of PTW with fixed parameters

dblaschke-LANL committed:
speed up the latest test

dblaschke-LANL commented:
the test I just added should cover everything we discussed; pls let me know if I missed anything

luiarthur committed:
moved numba from group -> optional dep

luiarthur committed:
remove numba as group dep
